packages/lesswrong/server/resolvers/userResolvers.ts

```diff
@@ -301,3 +301,138 @@ createPaginatedResolver({
+  const response = await fetch(url, {
+    headers: {
+      Authorization: `Bearer ${apiKey}`,
+    },
+  });
+
+  if (!response.ok) {
+    throw new Error(`Airtable API responded with status ${response.status} - ${response.statusText}`);
+  }
+
+  // The data returned by Airtable has a "records" array.
+  // Each "record" has a "fields" object containing the columns we need.
+  const data = await response.json();
+  const records = data.records || [];
+
+  const unpackArray = (value: any) => Array.isArray(value) ? value[0] : value;
+
+  // Transform "records" into our AirtableLeaderboardResultType format
+  return records.map((record: any) => {
+    const name = unpackArray(record.fields?.["Leaderboard Display Name"]) ?? "Unknown";
+    const amountStr = unpackArray(record.fields?.["Leaderboard Amount"]) ?? "";
+    return {
+      name,
+      leaderboardAmount: amountStr === "" ? undefined : parseInt(amountStr, 10),
+    };
+  });
+}
+
+addGraphQLSchema(`
+  type AirtableLeaderboardResult {
+    name: String!
+    leaderboardAmount: Int
+  }
+`)
+
+let airtableCache: {
+  data?: AirtableLeaderboardResultType[];
+  timestamp: number;
+} = {
+  data: undefined,
+  timestamp: 0,
+};
+
+// We'll keep track of whether we have an in-flight promise so that we don't trigger multiple fetches simultaneously.
+let inFlightPromise: Promise<AirtableLeaderboardResultType[]> | null = null;
+
+// We'll consider data stale if it's older than 10 minutes.
+const STALE_TIME_MS = 1000 * 60 * 10;
+
+defineQuery({
+  name: "AirtableLeaderboards",
+  resultType: "[AirtableLeaderboardResult!]!",
+  fn: async () => {
+    if (airtableCache.data) {
+      const isStale = Date.now() - airtableCache.timestamp > STALE_TIME_MS;
+
+      // If the data is stale but no fetch is happening yet, start a new background fetch.
+      if (isStale && !inFlightPromise) {
+        void (async () => {
+          try {
+            inFlightPromise = fetchAirtableRecords();
+            const freshData = await inFlightPromise;
+            airtableCache.data = freshData;
+            airtableCache.timestamp = Date.now();
+          } finally {
+            inFlightPromise = null;
+          }
+        })();
+      }
+
+      // Regardless of staleness, return the cached data (stale-while-revalidate).
+      return airtableCache.data;
+    }
+
+    // If we have no data yet:
+    if (inFlightPromise) {
+      // If we're already fetching it, just await that same promise
+      // so we don't fire off multiple fetches.
+      return await inFlightPromise;
+    } else {
+      // Start a new fetch and wait for it before returning.
+      inFlightPromise = (async () => {
+        try {
+          const freshData = await fetchAirtableRecords();
+          airtableCache.data = freshData;
+          airtableCache.timestamp = Date.now();
+          return freshData;
+        } finally {
+          inFlightPromise = null;
+        }
+      })();
```

Comment:
Does this need to be different from the `isStale && !inFlightPromise` case?  I don't see a reason why the `isStale && !inFlightPromise` case can't also just assign the top-level IIFE to `inFlightPromise`.  If so, seems good to pull this out to a top-level function in the module and just call it twice - in the stale case, you don't need to await the returned promise.